src/org/opendatakit/briefcase/export/CipherFactory.java

```diff
@@ -0,0 +1,119 @@
+import org.opendatakit.briefcase.model.CryptoException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handles the initialization of Cipher objects for the decryption of submission
+ * and media files.
+ */
+final class CipherFactory {
+  private static final Logger log = LoggerFactory.getLogger(CipherFactory.class);
+  private static final int IV_BYTE_LENGTH = 16;
+  private final SecretKeySpec symmetricKey;
+  private final byte[] ivSeedArray;
+  private int ivCounter = 0;
+
+  private CipherFactory(String instanceId, byte[] symmetricKeyBytes) throws CryptoException {
+    symmetricKey = new SecretKeySpec(symmetricKeyBytes, "AES/CFB/PKCS5Padding");
+    try {
+      MessageDigest md = MessageDigest.getInstance("MD5");
+      md.update(instanceId.getBytes("UTF-8"));
+      md.update(symmetricKeyBytes);
+      byte[] messageDigest = md.digest();
+      ivSeedArray = new byte[IV_BYTE_LENGTH];
+      for (int i = 0; i < IV_BYTE_LENGTH; ++i) {
+        ivSeedArray[i] = messageDigest[(i % messageDigest.length)];
+      }
+    } catch (NoSuchAlgorithmException | UnsupportedEncodingException e) {
+      String msg = "Error constructing ivSeedArray";
+      log.error(msg, e);
+      throw new CryptoException(msg + " Cause: " + e);
+    }
+  }
+
+  /**
+   * Factory that initializes a new {@link CipherFactory} for the given instance ID,
+   * encryption key and private key values.
+   *
+   * @throws CryptoException
+   */
+  static CipherFactory from(String instanceId, String base64EncryptedKey, PrivateKey privateKey) {
+    try {
+      Cipher pkCipher;
+      pkCipher = Cipher.getInstance("RSA/NONE/OAEPWithSHA256AndMGF1Padding");
+      pkCipher.init(Cipher.DECRYPT_MODE, privateKey);
+      byte[] encryptedSymmetricKey = Base64.decodeBase64(base64EncryptedKey);
+      byte[] decryptedKey = pkCipher.doFinal(encryptedSymmetricKey);
+      return new CipherFactory(instanceId, decryptedKey);
+    } catch (NoSuchPaddingException | NoSuchAlgorithmException | BadPaddingException | IllegalBlockSizeException | InvalidKeyException e) {
+      throw new CryptoException(e);
+    }
+  }
+
+  /**
+   * Factory that returns a {@link Cipher} suitable to decrypt cryptographic
+   * signatures.
+   *
+   * @throws CryptoException
+   */
+  static Cipher signatureDecrypter(PrivateKey privateKey) {
+    try {
+      Cipher pkCipher = Cipher.getInstance("RSA/NONE/OAEPWithSHA256AndMGF1Padding");
+      pkCipher.init(Cipher.DECRYPT_MODE, privateKey);
+      return pkCipher;
+    } catch (NoSuchPaddingException | NoSuchAlgorithmException | InvalidKeyException e) {
+      throw new CryptoException(e);
+    }
+  }
+
+  /**
+   * Return the next {@link Cipher} instance. This method has side-effects and will
+   * change the initialization vector, which will affect the next call to this method.
+   *
+   * @throws CryptoException
+   */
+  Cipher next() {
+    try {
+      ++ivSeedArray[ivCounter % ivSeedArray.length];
+      ++ivCounter;
```

Comment:
I’m assuming you didn’t write
`++ivSeedArray[ivCounter++ % ...` because you thought it would be too hard to understand. Some don’t know the difference between pre- and post-increment?

Reply:
Most of the code in this class is a mystery to me... :/

Reply:
Sorry, I thought you wrote that.